octodns/provider/base.py

```diff
@@ -34,21 +34,26 @@ def __init__(self, id, apply_disabled=False,
 
     def _process_desired_zone(self, desired):
         '''
-        An opportunity for providers to modify that desired zone records before
-        planning.
-
-        - Must do their work and then call super with the results of that work
-        - Must not modify the `desired` parameter or its records and should
-          make a copy of anything it's modifying
+        An opportunity for providers to modify the desired zone records before
+        planning. `desired` is a "shallow" copy, see `Zone.copy` for more
+        information
+
+        - Must call `super` at an appropriate point for their work, generally
+          that means as the final step of the method, returning the result of
+          the `super` call.
+        - May modify `desired` directly.
+        - Must not modify records directly, `record.copy` should be called,
```

Comment:
Do you think it might be worth it to make record objects "immutable" during this phase so that trying to set an attribute would result in an exception?

Reply:
That could be interesting. Will have to think about it/look into it.

Reply:
Changes required to do this were fairly involved. Bascially everything has to be accessed through `@property` w/o setters. Still like the idea of it, but definitely out of scope for now.